### PR TITLE
For Review - Twincam  Uptime file

### PR DESCRIPTION
**Pull request by iamianmullins**
As discussed, thanks Eric

### CODE REVIEW

**Comment by ericcurtin** [comment]
What you want to do here is something like:

```
diff --git a/meson.build b/meson.build
index ece28a2..99fedaa 100644
--- a/meson.build
+++ b/meson.build
@@ -17,7 +17,8 @@ twincam_sources = files([
     'src/event_loop.cpp',
     'src/twincam.cpp',
     'src/drm.cpp',
-    'src/kms_sink.cpp'
+    'src/kms_sink.cpp',
+    'src/uptime.cpp'
 ])

 twincam  = executable('twincam', twincam_sources, include_directories : incdir,
diff --git a/src/uptime.cpp b/src/uptime.cpp
new file mode 100644
index 0000000..1cd94f3
--- /dev/null
+++ b/src/uptime.cpp
@@ -0,0 +1,49 @@
+#include <fcntl.h>
+#include <locale.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+int uptime(float* up, float* elapsed) {
+  int fd;
+  const char* filename = "/proc/uptime";
+  char buf[32];
+  int n;
+  if ((fd = open(filename, O_RDONLY)) == -1) {
+    fputs("Error: /proc must be mounted\n", stderr);
+    fflush(NULL);
+    return 102;
+  }
+
+  lseek(fd, 0L, SEEK_SET);
+  if ((n = read(fd, buf, sizeof buf - 1)) < 0) {
+    perror(filename);
+    fflush(NULL);
+    return 103;
+  }
+
+  buf[n] = '\0';
+
+  char* savelocale = strdup(setlocale(LC_NUMERIC, NULL));
+  setlocale(LC_NUMERIC, "C");
+  if (sscanf(buf, "%f", up) < 1) {
+    setlocale(LC_NUMERIC, savelocale);
+    free(savelocale);
+    fputs("bad data in /proc/uptime\n", stderr);
+    return 104;
+  }
+
+  static float init_time = 0;
+  if (init_time) {
+    *elapsed = *up - init_time;
+  } else {
+    init_time = *up;
+    *elapsed = 0;
+  }
+
+  setlocale(LC_NUMERIC, savelocale);
+  free(savelocale);
+  return 0;
+}
+
diff --git a/src/uptime.h b/src/uptime.h
index 9c7939a..f18276d 100644
--- a/src/uptime.h
+++ b/src/uptime.h
@@ -1,51 +1,4 @@
-#include <fcntl.h>
-#include <locale.h>
-#include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
-#include <unistd.h>
-
-int uptime(float* up, float* elapsed) {
-  int fd;
-  const char* filename = "/proc/uptime";
-  char buf[32];
-  int n;
-  if ((fd = open(filename, O_RDONLY)) == -1) {
-    fputs("Error: /proc must be mounted\n", stderr);
-    fflush(NULL);
-    return 102;
-  }
-
-  lseek(fd, 0L, SEEK_SET);
-  if ((n = read(fd, buf, sizeof buf - 1)) < 0) {
-    perror(filename);
-    fflush(NULL);
-    return 103;
-  }
-
-  buf[n] = '\0';
-
-  char* savelocale = strdup(setlocale(LC_NUMERIC, NULL));
-  setlocale(LC_NUMERIC, "C");
-  if (sscanf(buf, "%f", up) < 1) {
-    setlocale(LC_NUMERIC, savelocale);
-    free(savelocale);
-    fputs("bad data in /proc/uptime\n", stderr);
-    return 104;
-  }
-
-  static float init_time = 0;
-  if (init_time) {
-    *elapsed = *up - init_time;
-  } else {
-    init_time = *up;
-    *elapsed = 0;
-  }
-
-  setlocale(LC_NUMERIC, savelocale);
-  free(savelocale);
-  return 0;
-}
+int uptime(float* up, float* elapsed);

 #define PRINT_UPTIME()                                                     \
    {                                                                       \
@@ -55,5 +8,4 @@ int uptime(float* up, float* elapsed) {
     printf("%.2f (%.2f), %s at %s:%d\n", up, elapsed, __PRETTY_FUNCTION__, \
            __FILE__, __LINE__);                                            \
   }
-
-
+
```

Then you should build fine

**Comment by ericcurtin** [comment]
Forgot the include guard (we use #pragma once for include guards), so the header file should look like this:

```
#pragma once

int uptime(float* up, float* elapsed);

#define PRINT_UPTIME()                                                     \
   {                                                                       \
    float up;                                                              \
    float elapsed;                                                         \
    uptime(&up, &elapsed);                                                 \
    printf("%.2f (%.2f), %s at %s:%d\n", up, elapsed, __PRETTY_FUNCTION__, \
           __FILE__, __LINE__);                                            \
  }
```

and the rest in the cpp file.

**Comment by iamianmullins** [comment]
That's great, thanks for the feedback @ericcurtin . I'll go ahead and implement what you have suggested above.

File: src/uptime.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#pragma once
```
Comment (ericcurtin) [inline src/uptime.h:1]: This looks fine

File: src/uptime.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    init_time = *up;
+    *elapsed = 0;
+  }
+
+  setlocale(LC_NUMERIC, savelocale);
+  free(savelocale);
+  return 0;
+}
+
+#define PRINT_UPTIME()                                                     \
```
Comment (ericcurtin) [inline src/uptime.cpp:50]: This should only be in one place in the header file, duplicate code...

File: meson.build [hunk 3]
```diff
@@ -17,7 +17,8 @@ twincam_sources = files([
     'src/event_loop.cpp',
```
Comment (ericcurtin) [inline meson.build:17]: This looks perfect

**Comment by ericcurtin** [comment]
Sample output:

```
38762.86 (0.00), int main(int, char**) at ../src/twincam.cpp:181
38762.97 (0.11), int CamApp::exec(int, char**) at ../src/twincam.cpp:77
38762.97 (0.11), int CamApp::run(int, char**) at ../src/twincam.cpp:91
38763.02 (0.16), KMSSink::KMSSink(const string&) at ../src/kms_sink.cpp:27
38763.05 (0.19), int KMSSink::configure(const libcamera::CameraConfiguration&) at ../src/kms_sink.cpp:101
38763.05 (0.19), int KMSSink::configurePipeline(const libcamera::PixelFormat&) at ../src/kms_sink.cpp:182
38763.05 (0.19), int KMSSink::selectPipeline(const libcamera::PixelFormat&) at ../src/kms_sink.cpp:126
Using KMS plane 31, CRTC 51, connector eDP-1 (95)
38763.05 (0.19), void KMSSink::mapBuffer(libcamera::FrameBuffer*) at ../src/kms_sink.cpp:66
38763.05 (0.19), void KMSSink::mapBuffer(libcamera::FrameBuffer*) at ../src/kms_sink.cpp:66
38763.05 (0.19), void KMSSink::mapBuffer(libcamera::FrameBuffer*) at ../src/kms_sink.cpp:66
38763.05 (0.19), void KMSSink::mapBuffer(libcamera::FrameBuffer*) at ../src/kms_sink.cpp:66
```

We probably want to add a few to CameraSession, we basically want to capture every functions start time, up until the first frame... We also need to add a command line argument to turn this on... Because in normal operation this would pollute the output...

**Comment by ericcurtin** [comment]
Example branch of command line parameter

https://github.com/ericcurtin/twincam/tree/uptime

**Comment by ericcurtin** [comment]
Pushed as 9d4e7d046ad75c7eb99dc32ad6fd1eb814798445

**Pull request closed without merging**